src/server/game/Chat/Hyperlinks.cpp

```diff
@@ -0,0 +1,404 @@
+                equal_with_len(info->SpellName[i], pos + (skillLen + 2), len - (skillLen + 2)))
+                return true;
+        }
+        return false;
+    }
+
+    static bool IsColorValid(SpellInfo const*, HyperlinkColor c)
+    {
+        return c == CHAT_LINK_COLOR_ENCHANT;
+    }
+};
+
+template <>
+struct LinkValidator<LinkTags::item>
+{
+    static bool IsTextValid(ItemLinkData const& data, char const* pos, size_t len)
+    {
+        ItemLocale const* locale = sObjectMgr->GetItemLocale(data.item->ItemId);
+
+        char const* const* randomSuffix = nullptr;
+        if (data.randomPropertyId < 0)
+        {
+            if (ItemRandomSuffixEntry const* suffixEntry = sItemRandomSuffixStore.LookupEntry(-data.randomPropertyId))
+                randomSuffix = suffixEntry->nameSuffix;
+            else
+                return false;
+        }
+        else if (data.randomPropertyId > 0)
+        {
+            if (ItemRandomPropertiesEntry const* propEntry = sItemRandomPropertiesStore.LookupEntry(data.randomPropertyId))
+                randomSuffix = propEntry->nameSuffix;
+            else
+                return false;
+        }
+
+        for (uint8 i = 0; i < TOTAL_LOCALES; ++i)
+        {
+            if (!locale && i != DEFAULT_LOCALE)
+                continue;
+            std::string const& name = (i == DEFAULT_LOCALE) ? data.item->Name1 : locale->Name[i];
+            if (name.empty())
+                continue;
+            if (randomSuffix)
+            {
+                if (len > name.length() + 1 &&
+                  (strncmp(name.c_str(), pos, name.length()) == 0) && 
+                  (*(pos + name.length()) == ' ') &&
+                  equal_with_len(randomSuffix[i], pos + name.length() + 1, len - name.length() - 1))
+                    return true;
+            }
+            else if (equal_with_len(name.c_str(), pos, len))
+                return true;
+        }
+        return false;
+    }
+
+    static bool IsColorValid(ItemLinkData const& data, HyperlinkColor c)
+    {
+        return c == ItemQualityColors[data.item->Quality];
+    }
+};
+
+template <>
+struct LinkValidator<LinkTags::quest>
+{
+    static bool IsTextValid(QuestLinkData const& data, char const* pos, size_t len)
+    {
+        QuestLocale const* locale = sObjectMgr->GetQuestLocale(data.quest->GetQuestId());
+        if (!locale)
+            return equal_with_len(data.quest->GetTitle().c_str(), pos, len);
+
+        for (uint8 i = 0; i < TOTAL_LOCALES; ++i)
+        {
+            std::string const& name = (i == DEFAULT_LOCALE) ? data.quest->GetTitle() : locale->Title[i];
+            if (name.empty())
+                continue;
+            if (equal_with_len(name.c_str(), pos, len))
+                return true;
+        }
+
+        return false;
+    }
+
+    static bool IsColorValid(QuestLinkData const&, HyperlinkColor)
+    {
+        return true; // this color will be overridden clientside anyway
```

Comment:
should at least validate if it falls in allowed color set

Reply:
the client completely ignores the color specified here and replaces it for the appropriate difficulty color code for the current character